README.md
# Miniature power tools for ECMAScript.

A collection of powerful utility functions that are generally too small to merit implementation as stand-alone modules.

- [Accurately discriminate variable types](#type) (not fooled by toStringTag overloading).
- [Create or add properties to objects](#typefixpd) using a compact notation for assigning specific attributes.
- [Create property-name aliases](#typefill), greatly simplifying support for syntax-flexible option parsing.
- [Merge properties and/or elements](#typemerge) from one container to another of potentially different type.
- [Copy data from any container to a Map](#typeasmap), permitting uniform has/get/set/delete syntax to be used.
- [Create pre-allocated ring-buffers](#ringbuffer) that support constant-time insertion and removal from *both* ends, making for very snappy queues.
- [Quickly assign a value to one of several contiguous ranges](#partition). Uses binary search techniques to support very large partition sets.
- [Define pointer-like index variables](#index) that support a post-increment syntax with a variety of useful automatic block-stitching options.
- [Return just the captures of regular expression matches](#capture), with optional default values and support for automatic type casting.
- [Iterate on objects](#foreach), with an optional accumulator, allowing complex expressions to be performed in a single iteration pass.
- [Trace the call parameters and result history](#callhistory) of a wrapped function to help with debugging.
- [Bounds-check array indices before use](#typeasindex), with default value support.
- [Wrap expensive object allocators](#pool) with a very simple pooler.

### Installation

`npm install @undercat/esutil`

Or clone this repository and include `util.js` directly in an HTML file. Although the module works in the browser, it will not benefit from module isolation if loaded directly, and the code required to use it differs somewhat:

_Node.js:_
```js
const util = require('@undercat/esutil'); // load the whole wrapper
util.Type(x);
const { Type } = require('@undercat/esutil'); // load a component (or several)
Type(x);
```
_Browser:_
```html
<script src='util.js'></script>
<script>
  const { Type } = $UC_esutil;
  Type(x);
</script>
```

The file itself is written as a CommonJS module, and when it is used in the browser it suffers from unavoidable load-order dependencies. Most of my other modules require esutils, so if you're going to use it directly, you should generally include it first.

Incidentally, it is impossible to write a single file that loads as both a CommonJS module and an ECMAScript module because **export** is a privileged keyword that cannot be conditionally executed or overloaded _and_ it throws if it is not satisfied or is encountered in a file not loaded as a module _and_ the error cannot be caught because the keyword is restricted to the global scope (and therefore cannot be wrapped in a `try/catch` block). If **export** failed with a non-fatal warning, like much of the rest of ECMAScript that deals with resources, there would be no problem writing actual modules that could load as CommonJS and ES6 simultaneously.

## API Description

<h3 id='type'>Type(<i>var</i>);</h3>

Returns a string describing the type of the supplied variable. This string is similar to the type tag that can be retrieved from an object using `Object.prototype.toString.call()`, but it is not wrapped in superfluous text and cannot be fooled by defining custom string tags on an object.

object|typeof|toString.call|Type()
-|-|-|-
null|object|[object Null]|Null
new Date()|object|[object Date]|Date
/re/|object|[object RegExp]|RegExp
async function(){}|function|[object AsyncFunction]|AsyncFunction
new ArrayBuffer(8)|object|[object ArrayBuffer]|ArrayBuffer
a = []; a[Symbol.toStringTag] = 'Bogus';|object|[object Bogus]|Array

---
<h3 id='typetag'>Type.tag(<i>var</i>);</h3>

Similar to `Object.prototype.toString.call()`, but reads directly from the relevant symbol on the object or its prototype. It does not wrap the value returned in `[object ...]`, preferring to return just the tag itself ('Map', 'Array', etc.). Use this when you <i>want</i> to be 'fooled' by toStringTag overloads.

---
<h3 id='typegetpd'>Type.getPD(<i>obj</i>);</h3>

Equivalent to `Object.getOwnPropertyDescriptors(obj)`.

---
<h3 id='typefixpd'>Type.fixPD(<i>source</i>, <i>attribute_default</i>, <i>source_filter</i>, [<i>container</i>]);<br>Type.create(<i>prototype</i>, <i>source</i>, <i>attribute_default</i>, <i>source_filter</i>, [<i>container</i>]);<br>Type.add(<i>target</i>, <i>source</i>, <i>attribute_default</i>, <i>source_filter</i>, [<i>container</i>]);</h3>

<dl>
<dt>prototype</dt><dd>The prototype to use (when creating a new object).</dd>
<dt>target</dt><dd>The target object (when adding properties).</dd>
<dt>source</dt><dd>An object that supplies the properties to assign. The keys of this object use a special syntax to support attribute overrides and property-name aliases (see below).</dd>
<dt>default_attributes</dt><dd>An octal digit that specifies the attributes to use when assigning new properties: 0 = no permissions, 1 = enumerable, 2 = writable, 4 = configurable, etc. The default is to make all properties enumerable+writable (3) except functions, which are made enumerable only (1). The setters/getters of aliases are always hidden, to keep from cluttering up object iteration.

\# | C | W | E
:-:|:-:|:-:|:-:
0 | - | - | -
1 | - | - | &check;
2 | - | &check; | -
3 | - | &check; | &check;
4 | &check; | - | -
5 | &check; | - | &check;
6 | &check; | &check; | -
7 | &check; | &check; | &check;

The default attributes can be overridden on a per-property basis by appending a colon and digit to the property-name-string (see example below).
</dd>
<dt>source_filter</dt><dd>Only those properties in <code>source</code> having these attributes will be considered when generating new target properties. Uses the same octal schema as <code>default_attributes</code>, above.</dd>
<dt>container</dt><dd>When using the '&#42;' syntax (see below), the backing object for hidden values can optionally be be reflected into the result, allowing those hidden values to be accessed directly. The <code>container</code> parameter gives the name to use for this reflection property. It defaults to '$', but can be set to <code>null</code> if you don&rsquo;t want the backing object to be reflected into the result at all.</dd>
<dt>(return value)</dt><dd>Type.fixPD() returns an object containing the 'fixed' property descriptors from <code>source</code>.<br>
Type.create() returns a new object created from the 'fixed' property descriptors.<br>
Type.add() returns <code>target</code> object with the 'fixed' properties added to it.</dd>
</dl>

The <code><b>source</b></code> object may contain special key-like-strings that specify...
<ul>
<li>what aliases, if any, should exist for a given property value,</li>
<li>what attributes the values and accessors should have (enumerable, writeable, configurable), and</li>
<li>whether values should be attached to the first name in the list, or &ldquo;hidden&rdquo; in a special backing object.</li>
</ul>

```js
const o = Type.create(Object.prototype, {
  'foo,alias:0': 123,
  'bar:5': function() { return this.foo; },
  'cat,dog,mammal,animal': true,
  'paws,mits*': { value: 4, set: "if (typeof v == 'number') $v = v;" },
  raw: 'no aliases'
});

console.log(Object.getOwnPropertyDescriptors(o));
{
  raw: {
    value: 'no aliases',
    writable: true,
    enumerable: true,
    configurable: false
  },
  foo: {
    value: 123,
    writable: false,
    enumerable: false,
    configurable: false
  },
  alias: {
    get: [Function: get],
    set: [Function: set],
    enumerable: false,
    configurable: false
  },
  bar: {
    value: [Function: bar:5],
    writable: false,
    enumerable: true,
    configurable: true
  },
  cat: {
    value: true,
    writable: true,
    enumerable: true,
    configurable: false
  },
  dog: {
    get: [Function: get],
    set: [Function: set],
    enumerable: true,
    configurable: false
  },
  mammal: {
    get: [Function: get],
    set: [Function: set],
    enumerable: true,
    configurable: false
  },
  animal: {
    get: [Function: get],
    set: [Function: set],
    enumerable: true,
    configurable: false
  },
  '$': {
    value: { paws: 4 },
    writable: false,
    enumerable: false,
    configurable: false
  },
  paws: {
    get: [Function (anonymous)],
    set: [Function: setter],
    enumerable: true,
    configurable: false
  },
  mits: {
    get: [Function (anonymous)],
    set: [Function: setter],
    enumerable: true,
    configurable: false
  }
}
```
...as you can see, the resulting object 'o' has four properties and four aliases:
- o.raw &#x21d2; an ordinary property that inherits the default attributes (enumerable+writable)
- o.foo &#x21d2; a 'hidden' number that is not enumerable, writable or configurable
- o.alias &#x21d2; an alias to <code>o.foo</code>
- o.bar &#x21d2; an enumerable+configurable function reference
- o.cat &#x21d2; a Boolean value
- o.dog | o.mammal | o.animal &#x21d2; aliases to <code>o.cat</code>
- o.paws | o.mits &#x21d2; aliases to <code>o.$.paws</code>

The first property name in a key-list is used to label the actual <i>value</i>; all the other names in a list are accessors (getters/setters) of that first property name. If the optional '&#42;' suffix is added to a name list, the value for all the names in the list is instead stored in a &ldquo;hidden&rdquo; object&mdash;the first name in the list winds up being just another accessor to that value. These '&#42;'-lists take objects as arguments, and the sub-properties of those objects are very similar to those used by <code>Object.defineProperty()</code>, except that they are restricted to...

<dl>
<dt>value</dt><dd>The value assigned to the hidden property value.</dd>
<dt>get</dt><dd>A getter for the hidden value. Although &ldquo;normal&rdquo; aliases are also implemented as accessors, those default accessors provide no way for the user to override their logic. The '&#42;' syntax allows the user to define fully-custom accessors, which then are used by <i>all</i> the names in the list. The raw value can still be accessed directly, if there is a reference to the backing object in the result, which there is, by default, in the non-enumerable property '$'. If a custom getter is not defined by the user, a default getter that returns the hidden value will be used (in other words, this sub-property is optional).</dd>
<dt>set</dt><dd>A setter for the hidden value. No default is provided, so if the user does not supply one, the resulting value will be read-only.</dd>

Getters and setters defined using the '&#42;' syntax need not actually be functions, although that is still an option. Simple strings giving the relevant logic of the getter or setter will suffice. In those strings, the value supplied to a setter will be identified by the variable <code>v</code>. The actual backing variable itself can be accessed through <code>$v</code>. If you write an actual function, no such translation will be performed, and you will have to use the &ldquo;real&rdquo; object path to the backing variable, so you might have to write something like, <code>'foo,bar*': { value: 123, set(x) { this.$.foo = x; } }</code> instead of <code>'foo,bar*': { value: 123, set: '$v = v;' }</code>. Note that the path can change if you supply a <code>container</code> parameter, which sets the identifier name for the backing object.

Performance-critical code may elect to bypass accessors and handle values directly, even for objects defined with the '&#42;' syntax. Ordinarily, the first name in the property name list is used to label the actual value, but if the '*' syntax is used, <i>all</i> the names will be aliases, and if you want direct access to the value, you will need to do it through the backing object reference, something like <code>myObject.$.myValue</code>. If you invoke <code>Type.fixPD()</code>, <code>Type.create()</code> or <code>Type.add()</code> with a null value for the <code>container</code> parameter, no reference to the backing object will appear in the resulting object, so there will be no &ldquo;fast path&rdquo; access to any such values.

The drawback to this property label processing scheme is that it precludes the possibility of defining &ldquo;ordinary&rdquo; property names containing commas, colons or asterisks. Since most people probably do not consider such property names to be &ldquo;ordinary&rdquo; anyway, this was considered to be an acceptable trade-off.

Another example will help clarify all this.

```js
const o = Type.create(Array.prototype, {
  // The first property uses a functional style setter, so the data must be explicitly referenced.
  'foo,a1*7': { value: 'text', set(x) { this.$$.foo = x + ' text'; }},
  'bar,a2*1': { value: 123 },
  // Using a string as a setter results in automatic source/target translation.
  'a3*': { value: new Date(), set: "$v = (typeof v == 'string') ? new Date(v) : v;" }
}, null, null, '$$');

console.log(Object.getOwnPropertyDescriptors(o));

-------
// The data object gets reflected into the result with the name '$$' instead of the default '$'
// It is never made enumerable, writeable or configurable. Moreover, it is sealed. (Not the values.)
// If null had also been used for the 'container' parameter, the data would not have been reflected
// and direct access would be impossible.
  '$$': {
    value: { foo: 'text', bar: 123, a3: '2020-04-30T14:17:25.866Z' },
    writable: false,
    enumerable: false,
    configurable: false
  },
// 'foo' and 'a1' are both accessors for the value in '#.$$.foo'
// The first name in the list is always used to label the value in the reflected data structure.
  foo: {
    get: [Function (anonymous)],
    set: [Function: set],
    enumerable: true,
    configurable: true  // defined with '7' attribute flag := all options, including configurable
  },
  a1: {
    get: [Function (anonymous)],
    set: [Function: set],
    enumerable: true,
    configurable: true
  },
// 'bar' and 'a2' are accessors for '#.$$.bar'
// Since no setter was declared for these, the value is read-only (unless read directly)
  bar: {
    get: [Function (anonymous)],
    set: undefined,
    enumerable: true,
    configurable: false  // attribute flag was '1', so only enumerable
  },
  a2: {
    get: [Function (anonymous)],
    set: undefined,
    enumerable: true,
    configurable: false
  },
// 'a3' accesses '#.$$.a3'
  a3: {
    get: [Function (anonymous)],
    set: [Function: setter],
    enumerable: true,
    configurable: false  // default attribute flag is '3' := enumerable+writeable
  }
// Note that accessors never have a 'writeable' attribute, so that flag value is irrelevant.
-------

console.log(o.foo);
// => 'text'

o.foo = 'hello'; console.log(o.a1);
// => 'hello'

o.a3 = '1980-10-01'; console.log(o.a3);
// => 1980-10-01T00:00:00.000Z
```

If you do not reflect the data object into the result (i.e., if you use <code>null</code> as the container parameter), then you <i>must</i> declare all getters and setters using string style, as there will not be any way to access the data with 'this'.
Since string-style getters and setters are translated at construction time, there is no need to use 'this' to reference any data.

---
<h3 id='typefill'>Type.fill(<i>target</i>, <i>source</i>, [<i>filter</i>]);</h3>

Assigns selected properties of <i>source</i>, which may be an Object or Map, to <i>target</i>, which must be an Object. Properties are selected either be being in the target object already (presumably with a different value), or by being present in a <i>filter</i> object, which may itself be an Array, Object, Map or Set. Only the keys of Maps and Objects are referenced in a <i>filter</i>.

If the target object has aliases defined on it (see above), multiple keys from <i>source</i> may map to the same actual value in <i>target</i>, in which case the last property encountered in <i>source</i> as it is being iterated will prevail. This ambiguity is actually quite useful for allowing options to be satisfied by multiple keywords, as the following example shows:

```js
const prog_opt = Type.create(Object.prototype, {
  'color,fgColor,foregroundColor': 'black',
  'bgColor,backgroundColor': 'white'
});
let user_opt1 = { color: 'red', bgColor: 'yellow', foo: 'purple' };
let user_opt2 = { fgColor: 'red', bgColor: 'yellow', bar: 'purple' };
let user_opt3 = { foregroundColor: 'red', backgroundColor: 'yellow' };

Type.fill(prog_opt, user_opt1);
Type.fill(prog_opt, user_opt2);
Type.fill(prog_opt, user_opt3);
```
In this example, <i>all</i> of the user option structures produce the <i>same</i> program option object. Essentially, the user can choose between label aliases in specifying a given option instead of being forced to use a canonical form. Any 'invalid' properties will be ignored.

---
<h3 id='typemerge'>Type.merge(<i>target</i>, <i>source</i>, <i>keys</i>, <i>props</i>);</h3>

<dl>
<dt>target</dt><dd>The container to merge elements into, either an Object, Array, Map or Set.</dd>
<dt>source</dt><dd>Supplies elements to transfer. It does not need to be the same type as the target.</dd>
<dt>keys<dt><dd>If the target is an Array or Set, then Object and Maps will not be able to transfer both keys and values to them simultaneously. If this flag is set, the keys from the Object or Map are transferred; if it is clear, the values are transferred.
<dt>props</dt><dd>If the source object is not an <i>actual</i> Object, any properties defined on the container (as opposed to elements <i>of</i> the container) will only be copied if this flag is set. Properties are always copied if the source is an actual Object, because it has no other elements to provide.</dd>
</dl>

```js
console.log(Type.merge(new Map([['foo',64]]), { bar: 23 })); // Map(2) { 'foo' => 64, 'bar' => 23 }
console.log(Type.merge([1,2,3], { foo: 4, bar: 6 })); // [ 1, 2, 3, 4, 6 ]
console.log(Type.merge([1,2,3], { foo: 4, bar: 6 }, true)); // [ 1, 2, 3, 'foo', 'bar' ]
console.log(Type.merge([1,2,3], Object.defineProperty([4,5,6], 'foo', { enumerable: true, value: 23 }), null, true));
// [ 1, 2, 3, 4, 5, 6, foo: 23 ]
```

---
<h3 id='typeasmap'>Type.asMap(<i>source</i>, [<i>recursive</i>]);</h3>

Copies all the elements from a _source_ Array, Map or Set, or copies all the enumerable, intrinsic ("own") properties from a _source_ Object, into a new stand-alone Map object. Useful for allowing code to be written for a single interface (namely, Map's) while accepting data from any container type. Arrays elements supply keys for the resulting Map entries, while their indices (i.e., order) supplies the values. Set elements are used as both the keys and values.

If the <i>recursive</i> flag is set to a Number or `true`, Type.asMap() will recursively convert containers in the appropriate number of object levels (or all of them) to Maps. Because Maps can use object instances, like Arrays and Sets, as actual keys, it would be difficult to distinguish between objects intended to be keys and objects intended to be collections on a level-by-level basis, so they are never recursively converted.

```js
console.log(Type.asMap({ foo: 12, bar:64 }));    // Map(2) { 'foo' => 12, 'bar' => 64 }
console.log(Type.asMap(['foo', 'bar']));         // Map(2) { 'foo' => 0, 'bar' => 1 }
console.log(Type.asMap(new Set(['foo','bar']))); // Map(2) { 'foo' => 'foo', 'bar' => 'bar' }
console.log(Type.asMap({ foo: 12, bar: { yes: 23, no: 34 }}, true));
// Map(2) { 'foo' => 12, Map(2) { 'yes' => 23, 'no' = 34 } }
```

---
<h3 id='typeasindex'>Type.asIndex(<i>object</i>, <i>elements</i>, <i>default_index</i>);</h3>

<dl>
<dt>object</dt><dd>Any object, which may or may not be covertible to a Number, and which may or may not be in-bounds.</dd>
<dt>elements</dt><dd>If <i>object</i> can be converted to a Number (or is a Number), it must lie between zero and this value. Default is 2<sup>30</sup>, to provide a safe margin on V8's index limit of &plusmn;2<sup>31</sup>.</dd>
<dt>default_index</dt><dd>If <i>object</i> cannot be converted to a Number, or is out-of-bounds, this index will be returned.</dd>
</dl>

```js
let a = [10,20,30,40,50], i = 45;
console.log(a[Type.asIndex(i, 5, 0)]); // 10
console.log(a[Type.asIndex(i - 41, 5, 0)]); // 50
```

---
<h3 id='index'>Index();<br>Index(<i>initial_value</i>, [mask]);<br>Index(<i>block_list</i>);<br>Index(<i>block_function</i>, [<i>...arguments</i>]);</h3>

If no argument is supplied, a function is returned that increments a counter from zero each time it is called (the default is to increment by one, unless an argument is supplied giving the increment value). It returns the <i>pre</i>-increment value. This allows post-increment expressions to be created in-line, like pointers in C++.

If a Number is supplied as the initial argument, that value is used to set the counter's initial value. If an optional integer mask is supplied, it will be ANDed with the counter after each operation...cheap modulus wrapping for rings.

If an Array is supplied, it may take two forms: `[block_size, base1, base2, base3, ...]` or `[[base1, length1], [base2, length2], ...]`. The first form uses a uniform block size built on any number of block bases; the second form supplies both the base and length for each block explicitly. Index() will thread blocks transparently, updating the counter when space in one block is exhausted with the next block base. If an increment is too large to fit in the remaining 'space' within a block, the next block of suitable size is returned.

If a Function is supplied as the initial argument, it is invoked both to supply the base and length of the initial block and each following block. The callback function is invoked with the following parameters:
```block_threader(null, null, ...threader_arguments)```
...to supply the first block, and...
```block_threader(block_base, block_length, ...threader_arguments)```
to supply all subsequent blocks. It should return an array `[block_base, block_length]` that describes the base and length of the next block, which will be supplied to it again, in turn, when another block is need (i.e., they are 'stable' and can be used as keys).

The object returned by Index() is itself a function. That function takes one optional parameter: the number by which to increment (or decrement, for negative numbers) the value of the counter. It returns the <i>unincremented</i> value, however, turning the function into a sort of 'pseudo-post-incremented-pointer' for pointerless languages, like ECMAScript.

There are also a few properties defined on the returned function-object:
- value — returns the current value of the internal index

CHANGELOG

v1.1 - 2020-04-30

  Added a feature to create "hidden properties" in objects produced by Type.create(),
  or extended with Type.add(). These hidden properties use accessors (getters/setters)
  to manipulate the actual data, which is held in a data object that need not, but can be,
  referenced in the returned object.

v1.0 - 2020-02-15

  Initial release.


validate.js
"use strict";
const { Type, Index, Pool, capture, callHistory, forEach, partition, RingBuffer } = require('./util.js'),
types = [undefined,null,true,12.3,123n,'hello',new Date(),/regex/,Symbol(),() => 'foo',function*(){},async function(){},{},[],new Map(),new Set(),
  new ArrayBuffer(64),new SharedArrayBuffer(64),new DataView(new ArrayBuffer(64)),new Uint32Array(4),
  new Error('test'),new SyntaxError('test2'),
  Object.defineProperties(new Set(['Cannot be fooled!']), { [Symbol.toStringTag]: { value: 'Bogus' }})];
let a, f, h, i = 0, o, p, r, t = {}, u;
function elog(s) { console.log('\x1b[34m' + s + '\x1b[m'); eval(s); }
function log(...a) { console.log(...a); }

log('\n\x1b[30;42m MEANINGFUL TYPE DISCRIMINATION \x1b[m');
for (const e of types) t[i++] = { 'typeof': typeof e, 'Symbol.toStringTag': Type.tag(e), 'Type()': Type(e) }
console.table(t);

log('\n\x1b[30;42m BOUNDS-CHECKED INDEX VALIDATION \x1b[m');
elog("t = ['first','second','third']; log(t[Type.asIndex('0')], t[Type.asIndex(99, 3, 1)], t[Type.asIndex('foo', 3, 2)]);");

log('\n\x1b[30;42m CREATE AN OBJECT WITH SPECIFIED ATTRIBUTES (default: enumerable for functions, enumerable+writable for everything else) \x1b[m');
elog("t = Type.create(Object.prototype, { foo: 88, 'bar,alias1,alias2:7': 99 }); log(Object.getOwnPropertyDescriptors(t)); log(t.bar, t.alias2 = 5, t.alias1, t.bar);");
log('\x1b[32m...you can hide property values, forcing public access through accessor aliases\x1b[m');
elog(`t = Type.create(Object.prototype, { 'foo,alias1*': { value: 99, set: "if (typeof v == 'number') $v = v;" }, 'bar,alias2*': { value: 'word', set: "if (typeof v == 'string') $v = v;" }});`);
elog(`log(Object.getOwnPropertyDescriptors(t));`);
elog(`log(t.foo, t.alias1, t.bar, t.alias2);`);
elog(`t.foo = 123; t.bar = 'changed'; log(t.foo, t.alias1, t.bar, t.alias2);`);
elog(`t.foo = 'invalid'; t.bar = 235; log(t.foo, t.alias1, t.bar, t.alias2);`);
log('\x1b[32m...but you can still get direct access to the values aliased this way via a hidden property\x1b[m');
elog(`log(t); log(t.$);`);
log('\x1b[32mYou can also add properties to an existing object\x1b[m');
elog("log(Object.getOwnPropertyDescriptors(Type.add({ foo: 23 }, { bar: 83, yadda: 'something' })));");
log('\x1b[32m...or merge elements from one container to another\x1b[m');
elog("log(Type.merge(new Map([['foo',64]]), { bar: 23 }));");
log('\x1b[32m...and with fill(), you can even populate objects with values from another object...\x1b[m');
elog("t = Type.create(Object.prototype, { 'color,foregroundColor': 'black', 'bgColor,backgroundColor': 'white' });");
elog("log(Type.fill(t, { foregroundColor: 'red', backgroundColor: 'yellow', foo: 'bar' }));");
log('\x1b[32m...allowing only valid properties through, and selecting from one of more aliases.\x1b[m');

log('\n\x1b[30;42m POINTER-LIKE VALUES ALLOW POST-INCREMENT EXPRESSIONS \x1b[m');
elog("t = Index(); log(t.value, t(4), t(6), t.value, t(35), t.value);");
log('\x1b[32m...with optional ring-buffer wrapping behavior...\x1b[m');
elog("t = Index(0, 0xff); log(t(64), t(64), t(64), t(64), t(64), t.value);");
log('\x1b[32m...or block-spanning behavior (first element is block length)...\x1b[m');
elog("t = Index([300, 0, 1000]); log(t(64), t(64), t(64), t(64), t(64), t.value);");
elog("t = Index([[1000, 128], [2000,256], [3000, 256]]); log(t(64), t(64), t(64), t(64), t(64), t(64), t(64), t.value);");
log('\x1b[32m...or a user-supplied block allocator function\x1b[m');
elog("a = [1000, 2000]; i = 0;\nt = Index(() => [a[i++], 256]); log(t(64), t(64), t(64), t(64), t(64), t.value);");

log('\n\x1b[30;42m POOL KEEPS TRACK OF IDLE OBJECTS THAT ARE EXPENSIVE TO RECREATE \x1b[m');
elog("f = function() { let r = 0; for (let i = 0; i < 100000000; ++i) r += i; return r; }; p = Pool(10, f);");
elog("console.time('uncached'); t = p(); console.timeEnd('uncached'); p.free(t); console.time('cached'); t = p(); console.timeEnd('cached');");

log('\n\x1b[30;42m RETURN ONLY THE CAPTURES OF REGULAR EXPRESSIONS, WITH DEFAULT VALUES \x1b[m');
elog("log(capture('abbbbddd', /(ab*)(c+)?/, 'ab', 'c'));");
log('\x1b[32m...if dates or numbers are supplied as a default, the corresponding field is coerced to that type\x1b[m');
elog("log(capture('12345 number 2000-01-01 date', /(\\d+) (number) ([^ ]*) (date)/, 999, 999, new Date(), new Date()));");

log('\n\x1b[30;42m UNIFORM ITERATION FOR ALL CONTAINER TYPES \x1b[m');
elog("t = {}; forEach({foo: 10, bar: 20}, (v, k, a, _, i) => t[i] = { key: k, value: v }); console.table(t);");
elog("t = {}; forEach(['foo', 'bar'], (v, k, a, _, i) => t[i] = { key: k, value: v }); console.table(t);");
elog("t = {}; forEach(new Map([['foo', 10], ['bar', 20]]), (v, k, a, _, i) => t[i] = { key: k, value: v }); console.table(t);");
elog("t = {}; forEach(new Set(['foo', 'bar']), (v, k, a, _, i) => t[i] = { key: k, value: v }); console.table(t);");
log('\x1b[32mStrings can be iterated by characters\x1b[m');
elog("t = []; forEach('\u262f balance', c => t.push(c)); log(t);");
log('\x1b[32mThe accumulator allows complex tests or extractions to be written as expressions---a superset of #some, #every and #reduce...\x1b[m');
elog("log(forEach('abcdefghijklm', (v, k, a) => { if (v > 'f') a.push(v); return a; }, []));");
log('\x1b[32mA finalizer can be used to convert statements to expressions\x1b[m');
elog("log(forEach(['foo','bar'], (v, k, a) => a += `<li>${v}</li>`, '', a => `<ul>${a}</ul>`));");
log('\x1b[32mProperties are iterated over even if they\'re defined on container objects, but can be filtered out by ID parameter, or by setting max_properties to zero.\x1b[m');
elog("t = Object.defineProperties(['foo','bar'], { yadda: { enumerable: true, value: 46 }, hidden: { value: 764 }});");
elog("forEach(t, (v, k, a, t, i) => console.log(i, k, v, t));");
elog("forEach(t, (v, k, a, t, i) => console.log(i, k, v, t), null, null, null, null, null, 0);");

log('\n\x1b[30;42m YOU CAN RECORD A FUNCTION\'S CALL HISTORY FOR LATER ANALYSIS \x1b[m');
elog("f = callHistory(s => 'result of ' + s, 2); // allow 2**2 elements of history, just to test");
elog("f('first call'); f('second call'); f('third call'); f('fourth call'); f('should wrap'); f('ooops!');");
elog("console.table(f.history);");
log('\x1b[32m...but it\'s not without a cost in efficiency\x1b[m');
elog("f = i => t += i + Math.random(); h = callHistory(i => t += i + Math.random());");
elog("t = 0; console.time('raw'); for (let i = 0; i < 1000000; ++i) f(i); console.timeEnd('raw');");
elog("t = 0; console.time('rec'); for (let i = 0; i < 1000000; ++i) h(i); console.timeEnd('rec');");

log('\n\x1b[30;42m SORT A NUMBER INTO ONE OF SEVERAL BINS (PARTITIONING) \x1b[m');
elog("t = []; for (const e of [0,24,25,26,49,50,51,74,75,76,99,100,101]) t.push([e, partition(e, [[25,0],[50,1],[75,2],[100,3]], 4)]); console.log(t);");

log('\n\x1b[30;42m RING BUFFERS HAVE CONSTANT INSERT/REMOVE TIME AT BOTH ENDS \x1b[m');
log('\x1b[32m>> 64 elements...\x1b[m');
elog("a = []; t = 0; console.time('array'); for (let j = 0; j < 1024; ++j) { for (let i = 0; i < 64; ++i) a.push(i); for (let i = 0; i < 64; ++i) t += a.pop(); } console.timeEnd('array');");
elog("r = RingBuffer(); u = 0; console.time('ring'); for (let j = 0; j < 1024; ++j) { for (let i = 0; i < 64; ++i) r.push(i); for (let i = 0; i < 64; ++i) u += r.pop(); } console.timeEnd('ring'); console.log(`${t} =? ${u}`);");
log('\x1b[32m>> 256 elements...\x1b[m');
elog("a = []; t = 0; console.time('array'); for (let j = 0; j < 1024; ++j) { for (let i = 0; i < 256; ++i) a.push(i); for (let i = 0; i < 256; ++i) t += a.pop(); } console.timeEnd('array');");
elog("r = RingBuffer(9); u = 0; console.time('ring'); for (let j = 0; j < 1024; ++j) { for (let i = 0; i < 256; ++i) r.push(i); for (let i = 0; i < 256; ++i) u += r.pop(); } console.timeEnd('ring'); console.log(`${t} =? ${u}`);");
log('\x1b[32m>> 1024 elements...\x1b[m');
elog("a = []; t = 0; console.time('array'); for (let j = 0; j < 1024; ++j) { for (let i = 0; i < 1024; ++i) a.push(i); for (let i = 0; i < 1024; ++i) t += a.pop(); } console.timeEnd('array');");
elog("r = RingBuffer(11); u = 0; console.time('ring'); for (let j = 0; j < 1024; ++j) { for (let i = 0; i < 1024; ++i) r.push(i); for (let i = 0; i < 1024; ++i) u += r.pop(); } console.timeEnd('ring'); console.log(`${t} =? ${u}`);");
log('\x1b[32m>> reverse, 64 elements...\x1b[m');
elog("a = []; t = 0; console.time('array'); for (let j = 0; j < 1024; ++j) { for (let i = 0; i < 64; ++i) a.unshift(i); for (let i = 0; i < 64; ++i) t += a.shift(); } console.timeEnd('array');");
elog("r = RingBuffer(); u = 0; console.time('ring'); for (let j = 0; j < 1024; ++j) { for (let i = 0; i < 64; ++i) r.queue(i); for (let i = 0; i < 64; ++i) u += r.unqueue(); } console.timeEnd('ring'); console.log(`${t} =? ${u}`);");
log('\x1b[32m>> reverse, 256 elements...\x1b[m');
elog("a = []; t = 0; console.time('array'); for (let j = 0; j < 1024; ++j) { for (let i = 0; i < 256; ++i) a.unshift(i); for (let i = 0; i < 256; ++i) t += a.shift(); } console.timeEnd('array');");
elog("r = RingBuffer(9); u = 0; console.time('ring'); for (let j = 0; j < 1024; ++j) { for (let i = 0; i < 256; ++i) r.queue(i); for (let i = 0; i < 256; ++i) u += r.unqueue(); } console.timeEnd('ring'); console.log(`${t} =? ${u}`);");
log('\x1b[32m>> reverse, 1024 elements...\x1b[m');
elog("a = []; t = 0; console.time('array'); for (let j = 0; j < 1024; ++j) { for (let i = 0; i < 1024; ++i) a.unshift(i); for (let i = 0; i < 1024; ++i) t += a.shift(); } console.timeEnd('array');");
elog("r = RingBuffer(11); u = 0; console.time('ring'); for (let j = 0; j < 1024; ++j) { for (let i = 0; i < 1024; ++i) r.queue(i); for (let i = 0; i < 1024; ++i) u += r.unqueue(); } console.timeEnd('ring'); console.log(`${t} =? ${u}`);");
log('\x1b[32m>> queue, 64 elements...\x1b[m');
elog("a = []; t = 0; console.time('array'); for (let j = 0; j < 1024; ++j) { for (let i = 0; i < 64; ++i) a.unshift(i); for (let i = 0; i < 64; ++i) t += a.pop(); } console.timeEnd('array');");
elog("r = RingBuffer(); u = 0; console.time('ring'); for (let j = 0; j < 1024; ++j) { for (let i = 0; i < 64; ++i) r.queue(i); for (let i = 0; i < 64; ++i) u += r.pop(); } console.timeEnd('ring'); console.log(`${t} =? ${u}`);");
log('\x1b[32m>> queue, 256 elements...\x1b[m');
elog("a = []; t = 0; console.time('array'); for (let j = 0; j < 1024; ++j) { for (let i = 0; i < 256; ++i) a.unshift(i); for (let i = 0; i < 256; ++i) t += a.pop(); } console.timeEnd('array');");
elog("r = RingBuffer(9); u = 0; console.time('ring'); for (let j = 0; j < 1024; ++j) { for (let i = 0; i < 256; ++i) r.queue(i); for (let i = 0; i < 256; ++i) u += r.pop(); } console.timeEnd('ring'); console.log(`${t} =? ${u}`);");
log('\x1b[32m>> queue, 1024 elements...\x1b[m');
elog("a = []; t = 0; console.time('array'); for (let j = 0; j < 1024; ++j) { for (let i = 0; i < 1024; ++i) a.unshift(i); for (let i = 0; i < 1024; ++i) t += a.pop(); } console.timeEnd('array');");
elog("r = RingBuffer(11); u = 0; console.time('ring'); for (let j = 0; j < 1024; ++j) { for (let i = 0; i < 1024; ++i) r.queue(i); for (let i = 0; i < 1024; ++i) u += r.pop(); } console.timeEnd('ring'); console.log(`${t} =? ${u}`);");
